File: VB/TextEditPopUp/RepositoryItemTextEditPopUp.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Data
Imports System.Windows.Forms
Imports System.Text.RegularExpressions
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors
Imports DevExpress.Utils.Win

Namespace TEPopUp
	<UserRepositoryItem("RegisterTextEditPopUp")> _
	Friend Class RepositoryItemTextEditPopUp
		Inherits RepositoryItemPopupContainerEdit
		Private _dataSource As Object
		Private lb As ListBoxControl
		Private dv As DataView
		Private _displayMember As String, editValue As String = ""
		Private splitEditValue(-1) As String
		Private separator As Char = ";"c

		Public Const TextEditPopUpName As String = "TextEditPopUp"
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return TextEditPopUpName
			End Get
		End Property

		Public Shared Sub RegisterTextEditPopUp()
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(TextEditPopUpName, GetType(TextEditPopUp), GetType(RepositoryItemTextEditPopUp), GetType(PopupContainerEditViewInfo), New ButtonEditPainter(), True))
		End Sub

		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			BeginUpdate()
			MyBase.Assign(item)
			Dim source As RepositoryItemTextEditPopUp = TryCast(item, RepositoryItemTextEditPopUp)
			If source Is Nothing Then
				Return
			End If
			Me.DataSource = source.DataSource
			Me.DisplayMember = source.DisplayMember
			EndUpdate()
		End Sub

		Shared Sub New()
			RegisterTextEditPopUp()
		End Sub
		Public Sub New()
			MyBase.New()
			' Create PopupContainerControl with a ListBoxControl and bind it
			' to the editor
			lb = New ListBoxControl()
			lb.Dock = System.Windows.Forms.DockStyle.Fill
			AddHandler lb.Click, AddressOf lb_Click

			Dim pc As New PopupContainerControl()
			pc.Controls.Add(lb)

			Me.PopupControl = pc
			Me.TextEditStyle = DevExpress.XtraEditors.Controls.TextEditStyles.Standard
			AddHandler EditValueChanging, AddressOf RepositoryItemTextEditPopUp_EditValueChanging
			AddHandler KeyDown, AddressOf RepositoryItemTextEditPopUp_KeyDown
		End Sub

		Private Sub RepositoryItemTextEditPopUp_EditValueChanging(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.ChangingEventArgs)
			If e.NewValue IsNot Nothing Then
				editValue = e.NewValue.ToString()
			End If

			splitEditValue = editValue.Split(separator)

			If editValue.EndsWith(";") Then
				HidePopupHint(sender)
			End If
			ShowPopupHint(sender, splitEditValue(splitEditValue.Length - 1).TrimStart())
		End Sub

		Protected Overrides Sub OnOwnerEditChanged()
			AddHandler KeyDown, AddressOf RepositoryItemTextEditPopUp_KeyDown
			AddHandler OwnerEdit.Validating, AddressOf OwnerEdit_Validated
			MyBase.OnOwnerEditChanged()
		End Sub

		Private Sub OwnerEdit_Validated(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs)
			If (TryCast(sender, TextEditPopUp)).EditValue IsNot Nothing Then
				UpdateDataSource(Me.OwnerEdit.EditValue.ToString())
			End If
		End Sub

		Private Sub RepositoryItemTextEditPopUp_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
			If Not(TryCast(sender, TextEditPopUp)).IsPopupOpen Then
				Return
			End If
			PopupNavigation(sender, e.KeyValue)
			e.Handled = True
		End Sub

		Private Sub PopupNavigation(ByVal sender As Object, ByVal keyValue As Integer)
			Select Case keyValue
				Case 38
					' Arrow up key
					If lb.SelectedIndex > 0 Then
						lb.SelectedIndex -= 1
					End If
				Case 40
					' Arrow down key
					lb.SelectedIndex += 1
				Case 13
					' Enter key
					AppendToEditValue(sender, lb.Text)
					HidePopupHint(sender)
				Case 27
					' ESC key
					HidePopupHint(sender)
			End Select
		End Sub

		Private Sub lb_Click(ByVal sender As Object, ByVal e As EventArgs)
			If TypeOf sender Is ListBoxControl Then
				AppendToEditValue((TryCast((TryCast(sender, Control)).Parent, PopupContainerControl)).OwnerEdit, lb.Text)
				HidePopupHint((TryCast((TryCast(sender, Control)).Parent, PopupContainerControl)).OwnerEdit)
				Return
			End If
			AppendToEditValue(sender, lb.Text)
			HidePopupHint(sender)
		End Sub

		' Skip default button creation
		Public Overrides Sub CreateDefaultButton()
		End Sub

		Private Sub ShowPopupHint(ByVal sender As Object, ByVal word As String)
			If word.Length = 0 Then
				Return
			End If

			If SetDataViewFilter(word) Then
				TryCast(sender, TextEditPopUp).ShowPopup()
				TryCast(sender, TextEditPopUp).Focus()
				TryCast(sender, TextEditPopUp).SelectionStart = editValue.Length
			Else
				HidePopupHint(sender)
			End If
		End Sub

		Private Function SetDataViewFilter(ByVal word As String) As Boolean
			If DisplayMember Is Nothing OrElse dv Is Nothing Then
				Return False
			End If
			If word.Contains("'") Then
				Return False
			End If

			dv.RowFilter = "[" & DisplayMember & "] LIKE '" & word & "*'"
			Return If(dv.Count > 0, True, False)
		End Function

		Private Sub HidePopupHint(ByVal sender As Object)
			TryCast(sender, TextEditPopUp).ClosePopup()
			lb.SelectedIndex = -1
			TryCast(sender, TextEditPopUp).SelectionStart = editValue.Length
		End Sub

		Private Sub UpdateDataSource(ByVal value As String)
			If value.Length = 0 OrElse Me.DataSource Is Nothing Then
				Return
			End If

			Dim word As String
			For Each line As String In splitEditValue
				word = line.TrimStart()
				If word.Length = 0 Then
					Continue For
				End If

				If (Not ValueExistsInDataSource(word)) Then
					AddValueToDataSource(word)
				End If
			Next line
		End Sub

		Private Function ValueExistsInDataSource(ByVal word As String) As Boolean
			For Each dr As DataRow In (TryCast(Me.DataSource, DataTable)).Rows
				Dim str As String = dr(DisplayMember).ToString()
				If String.Compare(dr(DisplayMember).ToString(), word, True) = 0 Then
					Return True
				End If
			Next dr
			Return False
		End Function

		Private Sub AddValueToDataSource(ByVal value As String)
			Dim row As DataRow = (TryCast(Me.DataSource, DataTable)).NewRow()
			row(DisplayMember) = value
			TryCast(Me.DataSource, DataTable).Rows.Add(row)
		End Sub

		Private Sub AppendToEditValue(ByVal sender As Object, ByVal word As String)
			Dim value As String = ""
			For i As Integer = 0 To splitEditValue.Length - 2
				value &= splitEditValue(i).TrimStart() & "; "
			Next i

			value &= word
			If TypeOf sender Is TextEditPopUp Then
				TryCast(sender, TextEditPopUp).Text = value
			Else
				TryCast((TryCast(sender, Control)).Parent, PopupContainerControl).OwnerEdit.Text = value
			End If
		End Sub

		Public Overridable Property DataSource() As Object
			Get
				Return _dataSource
			End Get
			Set(ByVal value As Object)
				If value Is DataSource Then
					Return
				End If
				If value IsNot Nothing AndAlso DataSource IsNot Nothing AndAlso DataSource.Equals(value) Then
					Return
				End If
				_dataSource = value
				OnDataSourceChanged()
			End Set
		End Property

		Private Sub OnDataSourceChanged()
			If Me.DataSource Is Nothing Then
				Return
			End If
			dv = New DataView(TryCast(Me.DataSource, DataTable))
			lb.DataSource = dv
		End Sub

		Public Property DisplayMember() As String
			Get
				Return _displayMember
			End Get
			Set(ByVal value As String)
				If value Is Nothing OrElse value = DisplayMember Then
					Return
				End If
				_displayMember = value
				OnDisplayMemberChanged()
			End Set
		End Property

		Private Sub OnDisplayMemberChanged()
			lb.DisplayMember = _displayMember
		End Sub
	End Class
End Namespace

File: CS/TextEditPopUp/RepositoryItemTextEditPopUp.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Data;
using System.Windows.Forms;
using System.Text.RegularExpressions;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors;
using DevExpress.Utils.Win;

namespace TEPopUp
{
    [UserRepositoryItem("RegisterTextEditPopUp")]
    class RepositoryItemTextEditPopUp : RepositoryItemPopupContainerEdit
    {
        object _dataSource;
        ListBoxControl lb;
        DataView dv;
        string _displayMember, editValue = "";
        string[] splitEditValue = new string[0];
        char separator = ';';

        public const string TextEditPopUpName = "TextEditPopUp";
        public override string EditorTypeName { get { return TextEditPopUpName; } }

        public static void RegisterTextEditPopUp()
        {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(
                TextEditPopUpName, typeof(TextEditPopUp), typeof(RepositoryItemTextEditPopUp),
                typeof(PopupContainerEditViewInfo), new ButtonEditPainter(), true));
        }
        
        public override void Assign(RepositoryItem item)
        {
            BeginUpdate();
            base.Assign(item);
            RepositoryItemTextEditPopUp source = item as RepositoryItemTextEditPopUp;
            if (source == null) return;
            this.DataSource = source.DataSource;
            this.DisplayMember = source.DisplayMember;
            EndUpdate();
        }

        static RepositoryItemTextEditPopUp() { RegisterTextEditPopUp(); }
        public RepositoryItemTextEditPopUp()
            : base()
        {
            // Create PopupContainerControl with a ListBoxControl and bind it
            // to the editor
            lb = new ListBoxControl();
            lb.Dock = System.Windows.Forms.DockStyle.Fill;
            lb.Click += new EventHandler(lb_Click);

            PopupContainerControl pc = new PopupContainerControl();
            pc.Controls.Add(lb);

            this.PopupControl = pc;
            this.TextEditStyle = DevExpress.XtraEditors.Controls.TextEditStyles.Standard;
            this.EditValueChanging += new DevExpress.XtraEditors.Controls.ChangingEventHandler(RepositoryItemTextEditPopUp_EditValueChanging);
            this.KeyDown += new KeyEventHandler(RepositoryItemTextEditPopUp_KeyDown);
        }

        void RepositoryItemTextEditPopUp_EditValueChanging(object sender, DevExpress.XtraEditors.Controls.ChangingEventArgs e)
        {
            if (e.NewValue != null)
                editValue = e.NewValue.ToString();
            
            splitEditValue = editValue.Split(separator);

            if (editValue.EndsWith(";"))
                HidePopupHint(sender);
            ShowPopupHint(sender, splitEditValue[splitEditValue.Length - 1].TrimStart());
        }

        protected override void OnOwnerEditChanged()
        {
            this.KeyDown += new KeyEventHandler(RepositoryItemTextEditPopUp_KeyDown);
            this.OwnerEdit.Validating += new System.ComponentModel.CancelEventHandler(OwnerEdit_Validated);
            base.OnOwnerEditChanged();
        }

        void OwnerEdit_Validated(object sender, System.ComponentModel.CancelEventArgs e)
        {
            if ((sender as TextEditPopUp).EditValue != null)
                UpdateDataSource(this.OwnerEdit.EditValue.ToString());
        }

        void RepositoryItemTextEditPopUp_KeyDown(object sender, KeyEventArgs e)
        {
            if (!(sender as TextEditPopUp).IsPopupOpen) return;
            PopupNavigation(sender, e.KeyValue);
            e.Handled = true;
        }

        void PopupNavigation(object sender, int keyValue)
        {
            switch (keyValue)
            {
                case 38:
                    // Arrow up key
                    if (lb.SelectedIndex > 0)
                        lb.SelectedIndex--;
                    break;
                case 40:
                    // Arrow down key
                    lb.SelectedIndex++;
                    break;
                case 13:
                    // Enter key
                    AppendToEditValue(sender, lb.Text);
                    HidePopupHint(sender);
                    break;
                case 27:
                    // ESC key
                    HidePopupHint(sender);
                    break;
            }
        }

        void lb_Click(object sender, EventArgs e)
        {
            if (sender is ListBoxControl)
            {
                AppendToEditValue(((sender as Control).Parent as PopupContainerControl).OwnerEdit, lb.Text);
                HidePopupHint(((sender as Control).Parent as PopupContainerControl).OwnerEdit);
                return;
            }
            AppendToEditValue(sender, lb.Text);
            HidePopupHint(sender);
        }
        
        // Skip default button creation
        public override void CreateDefaultButton() { }

        void ShowPopupHint(object sender, string word)
        {
            if (word.Length == 0) return;
            
            if (SetDataViewFilter(word))
            {
                (sender as TextEditPopUp).ShowPopup();
                (sender as TextEditPopUp).Focus();
                (sender as TextEditPopUp).SelectionStart = editValue.Length;
            }
            else
                HidePopupHint(sender);
        }

        bool SetDataViewFilter(string word)
        {
            if (DisplayMember == null || dv == null) return false;
            if (word.Contains("'")) return false;

            dv.RowFilter = "[" + DisplayMember + "] LIKE '" + word + "*'";
            return dv.Count > 0 ? true : false;
        }

        void HidePopupHint(object sender)
        {
            (sender as TextEditPopUp).ClosePopup();
            lb.SelectedIndex = -1;
            (sender as TextEditPopUp).SelectionStart = editValue.Length;
        }

        void UpdateDataSource(string value)
        {
            if (value.Length == 0 || this.DataSource == null) return;
            
            string word;
            foreach (string line in splitEditValue)
            {
                word = line.TrimStart();
                if (word.Length == 0) continue;

                if (!ValueExistsInDataSource(word))
                    AddValueToDataSource(word);
            }
        }

        bool ValueExistsInDataSource(string word)
        {
            foreach (DataRow dr in (this.DataSource as DataTable).Rows)
            {
                string str = dr[DisplayMember].ToString();
                if (String.Compare(dr[DisplayMember].ToString(), word, true) == 0)
                    return true;
            }
            return false;
        }

        void AddValueToDataSource(string value)
        {
            DataRow row = (this.DataSource as DataTable).NewRow();
            row[DisplayMember] = value;
            (this.DataSource as DataTable).Rows.Add(row);
        }

        void AppendToEditValue(object sender, string word)
        {
            string value = "";
            for (int i = 0; i < splitEditValue.Length - 1; i++)
                value += splitEditValue[i].TrimStart() + "; ";
            
            value += word;
            if (sender is TextEditPopUp)
                (sender as TextEditPopUp).Text = value;
            else
                ((sender as Control).Parent as PopupContainerControl).OwnerEdit.Text = value;
        }

        public virtual object DataSource
        {
            get { return _dataSource; }
            set {
                if (value == DataSource) return;
                if (value != null && DataSource != null && DataSource.Equals(value)) return;
                _dataSource = value;
                OnDataSourceChanged();
            }
        }

        void OnDataSourceChanged()
        {
            if (this.DataSource == null) return;
            dv = new DataView(this.DataSource as DataTable);
            lb.DataSource = dv;
        }

        public string DisplayMember 
        {
            get { return _displayMember; }
            set {
                if (value == null || value == DisplayMember) return;
                _displayMember = value;
                OnDisplayMemberChanged();
            }
        }

        void OnDisplayMemberChanged()
        { lb.DisplayMember = _displayMember; }
    }
}
